### PR TITLE
Segfault on policy handling

### LINKED ISSUE #642: segfault in kubearmor during policy handling
## Bug Report

During policy apply, kubearmor got into CrashLoopBackOff. After removing the policy the kubearmor got back up normally.

<details>
  <summary>Policy causing segfault</summary>

[cartservice.yaml.txt](https://github.com/kubearmor/KubeArmor/files/8193057/cartservice.yaml.txt) save this text file as yaml.

```yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: autopol-system-1760649700
  namespace: default
spec:
  severity: 1
  selector:
    matchLabels:
      app: cartservice
  process:
    matchPaths:
    - path: /app/cartservice
    - path: /bin/grpc_health_probe
    - path: /sbin/su-exec
    - path: /usr/bin/find
      fromSource:
      - path: /usr/local/bin/docker-entrypoint.sh
    - path: /usr/bin/id
      fromSource:
      - path: /usr/local/bin/docker-entrypoint.sh
    - path: /usr/local/bin/docker-entrypoint.sh
  file:
    matchPaths:
    - path: /cpu.cfs_period_us
      fromSource:
      - path: /app/cartservice
    - path: /cpu.cfs_quota_us
      fromSource:
      - path: /app/cartservice
    - path: /cpu
      fromSource:
      - path: /app/cartservice
    - path: /data
      fromSource:
      - path: /usr/bin/find
    - path: /dev/urandom
      fromSource:
      - path: /app/cartservice
    - path: /etc/group
      fromSource:
      - path: /bin/grpc_health_probe
      - path: /usr/local/bin/docker-entrypoint.sh
    - path: /etc/os-release
      fromSource:
      - path: /app/cartservice
    - path: /etc/passwd
      fromSource:
      - path: /bin/grpc_health_probe
      - path: /usr/bin/find
      - path: /usr/bin/id
      - path: /usr/local/bin/docker-entrypoint.sh
    - path: /lib/libz.so.1.2.11
      fromSource:
      - path: /app/cartservice
    - path: /memory/memory.limit_in_bytes
      fromSource:
      - path: /app/cartservice
    - path: /memory/memory.stat
      fromSource:
      - path: /app/cartservice
    - path: /usr/lib/libgcc_s.so.1
      fromSource:
      - path: /app/cartservice
    - path: /cpuacct/cpu.cfs_period_us
      fromSource:
      - path: /app/cartservice
    - path: /cpuacct/cpu.cfs_quota_us
      fromSource:
      - path: /app/cartservice
    matchDirectories:
    - dir: /proc/
      fromSource:
      - path: /app/cartservice
      - path: /bin/grpc_health_probe
    - dir: /usr/lib/
      fromSource:
      - path: /app/cartservice
    - dir: /sys/
      fromSource:
      - path: /app/cartservice
      - path: /bin/grpc_health_probe
  network:
    matchProtocols:
    - protocol: tcp
      fromSource:
      - path: /.NET
      - path: /app/cartservice
      - path: /bin/grpc_health_probe
      - path: /usr/local/bin/docker-entrypoint.sh
      - path: /DefaultSocketMa
      - path: /cartservice
      - path: /grpc_health_pro
    - protocol: udp
      fromSource:
      - path: /.NET
      - path: /cartservice
  action: Allow
---
```
</details>

<details>
  <summary>Segfault details</summary>

```
2022-03-06 16:36:54.084504	INFO	Added cartservice-5dff477f54-cpzrk into the pod list of the AppArmor profile (kubearmor-default-server, 8)
2022-03-06 16:36:54.084580	INFO	Detected a Pod (added/default/cartservice-5dff477f54-cpzrk)
panic: runtime error: slice bounds out of range [:-1]

goroutine 1920 [running]:
github.com/kubearmor/KubeArmor/KubeArmor/common.RemoveStringElement(...)
	/usr/src/KubeArmor/KubeArmor/common/common.go:34
github.com/kubearmor/KubeArmor/KubeArmor/enforcer.(*AppArmorEnforcer).ResolvedProcessWhiteListConflicts(0xc00042dc4a, 0xc000d34178, 0xc000a7e480, 0xc000d341f0)
	/usr/src/KubeArmor/KubeArmor/enforcer/appArmorProfile.go:32 +0x485
github.com/kubearmor/KubeArmor/KubeArmor/enforcer.(*AppArmorEnforcer).GenerateProfileBody(0xc000969470, {0xc000503d40, 0x1, 0xc000308dc8})
	/usr/src/KubeArmor/KubeArmor/enforcer/appArmorProfile.go:1303 +0x1d0
github.com/kubearmor/KubeArmor/KubeArmor/enforcer.(*AppArmorEnforcer).GenerateAppArmorProfile(0x0, {0xc000308dc8, 0x18}, {0xc000503d40, 0x1, 0x1})
	/usr/src/KubeArmor/KubeArmor/enforcer/appArmorProfile.go:1511 +0x130
github.com/kubearmor/KubeArmor/KubeArmor/enforcer.(*AppArmorEnforcer).UpdateAppArmorProfile(0xc0002566c0, {{0xc000b1ab72, 0x7}, {0xc0002c28c0, 0x1c}, 0xc000e1f590, {0xc000368680, 0x2, 0x2}, {0xc00052b2e0, ...}, ...}, ...)
	/usr/src/KubeArmor/KubeArmor/enforcer/appArmorEnforcer.go:451 +0x4e
github.com/kubearmor/KubeArmor/KubeArmor/enforcer.(*AppArmorEnforcer).UpdateSecurityPolicies(0xc0002566c0, {{0xc000b1ab72, 0x7}, {0xc0002c28c0, 0x1c}, 0xc000e1f590, {0xc000368680, 0x2, 0x2}, {0xc00052b2e0, ...}, ...})
	/usr/src/KubeArmor/KubeArmor/enforcer/appArmorEnforcer.go:513 +0x1c5
github.com/kubearmor/KubeArmor/KubeArmor/enforcer.(*RuntimeEnforcer).UpdateSecurityPolicies(0xc0002a29a0, {{0xc000b1ab72, 0x7}, {0xc0002c28c0, 0x1c}, 0xc000e1f590, {0xc000368680, 0x2, 0x2}, {0xc00052b2e0, ...}, ...})
	/usr/src/KubeArmor/KubeArmor/enforcer/runtimeEnforcer.go:139 +0x78
github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).UpdateEndPointWithPod(0xc000580b60, {0xc000b1ab68, 0x5}, {0xc000e3dec0, 0xc000e1f3e0, 0xc000e1f410, 0xc000e1f4a0, 0xc000e1f500})
	/usr/src/KubeArmor/KubeArmor/core/kubeUpdate.go:240 +0x23f8
github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).WatchK8sPods(0xc000580b60)
	/usr/src/KubeArmor/KubeArmor/core/kubeUpdate.go:699 +0x2070
created by github.com/kubearmor/KubeArmor/KubeArmor/core.KubeArmor
	/usr/src/KubeArmor/KubeArmor/core/kubeArmor.go:502 +0x905
```
</details>

Attached is the karmor-sysdump tgz.
[karmor-sysdump-Sun Mar  6 22:15:46 IST 2022.zip](https://github.com/kubearmor/KubeArmor/files/8193041/karmor-sysdump-Sun.Mar.6.22.15.46.IST.2022.zip)

**General Information**

- Environment description (GKE, VM-Kubeadm, vagrant-dev-env, minikube, microk8s, ...)
I tried in k3s env, but should happen in any k8s env.

- Kernel version (run `uname -a`)
any

- Orchestration system version in use (e.g. `kubectl version`, ...)
```
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:41:01Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.6+k3s1", GitCommit:"3228d9cb9a4727d48f60de4f1ab472f7c50df904", GitTreeState:"clean", BuildDate:"2022-01-25T01:27:44Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}
```
- Link to relevant artifacts (policies, deployments scripts, ...)
* Used [google microservice demo](https://github.com/GoogleCloudPlatform/microservices-demo)
* The policy causing segfault is attached above.

- Target containers/pods
* kubearmor daemonset
* kubearmor-relay
Both go into CrashLoopBackOff

**To Reproduce**

1. Install kubearmor in the sample k8s env (see [deployment-guide](https://github.com/kubearmor/KubeArmor/blob/main/getting-started/deployment_guide.md))
2. `kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/microservices-demo/main/release/kubernetes-manifests.yaml`
3. `kubectl apply -f cartservice.yaml`

**Expected behavior**

Kubearmor should be able to apply the policy without any issues (without any segfaults).

### PR DESCRIPTION
**Pull request by seswarrajan**
Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

### CODE REVIEW

File: KubeArmor/enforcer/appArmorProfile.go [hunk 1]
```diff
@@ -23,8 +23,8 @@ func (ae *AppArmorEnforcer) ResolvedProcessWhiteListConflicts(processWhiteList *
 	copy(prunedProcessWhiteList, *processWhiteList)
 	numOfRemovedElements := 0
 
-	for index, line := range *processWhiteList {
-		for source := range fromSources {
+	for source := range fromSources {
```
Comment (nyrahul) [inline KubeArmor/enforcer/appArmorProfile.go]: I am not sure how this fixes segfault issue! ... is `processWhiteList` nil when there are no fromSources? Such that if we swap the for conditions, the fromSources are checked first and the *processWhiteList check is never hit?

Comment (seswarrajan) [reply KubeArmor/enforcer/appArmorProfile.go]: @nyrahul  `processWhiteList` is not nil. The issue is that in source, the second time, the index is not incremented. That index should increment in second loop. 

`prunedProcessWhiteList = kl.RemoveStringElement(prunedProcessWhiteList, index-numOfRemovedElements)`

In this case, in second loop, the second argument is 0-1, which is -1 for a string slice. This causes array index out of bounds.

Comment (nam-jaehyun) [reply KubeArmor/enforcer/appArmorProfile.go]: I don't still get the point. can you explain the problem a bit more?

**Comment by seswarrajan** [comment]
Kindly ignore this commit as I am facing similar issue in other scenario.

File: KubeArmor/enforcer/appArmorProfile.go [hunk 2]
```diff
@@ -24,13 +24,17 @@ func (ae *AppArmorEnforcer) ResolvedProcessWhiteListConflicts(processWhiteList *
...
 		for source := range fromSources {
 			if strings.Contains(line, source) {
 				*fusionProcessWhiteList = append(*fusionProcessWhiteList, source)
 
-				// remove line from WhiteList
-				prunedProcessWhiteList = kl.RemoveStringElement(prunedProcessWhiteList, index-numOfRemovedElements)
-				numOfRemovedElements = numOfRemovedElements + 1
+				if !processed {
+					// remove line from WhiteList
+					prunedProcessWhiteList = kl.RemoveStringElement(prunedProcessWhiteList, index-numOfRemovedElements)
```
Comment (nyrahul) [inline KubeArmor/enforcer/appArmorProfile.go]: @seswarrajan, I am still not clear as to what was causing the segfault here?
This PR adds a check to call `kl.RemoveStringElement()` just once ... but even if it is called multiple times, how would it result in a segfault?

File: KubeArmor/enforcer/appArmorProfile.go [hunk 3]
```diff
@@ -25,12 +25,13 @@ func (ae *AppArmorEnforcer) ResolvedProcessWhiteListConflicts(processWhiteList *
 
 	for index, line := range *processWhiteList {
 		for source := range fromSources {
-			if strings.Contains(line, source) {
+			if strings.Contains(line, source) && line[:len(source)] == source {
```
Comment (nyrahul) [inline KubeArmor/enforcer/appArmorProfile.go]: I dont think this check is correct.

If I have line=`/app/cartservice ix` and source=`/app` then the check will pass!

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/enforcer/appArmorProfile.go (hunk 3 above)
Comment (seswarrajan) [reply KubeArmor/enforcer/appArmorProfile.go]: Understood Rahul. Let me check again.

**Review by nam-jaehyun — APPROVED** [review]
good to merge it

**Review by nyrahul — APPROVED** [review]

**Pull request merged**